client/client.go

```diff
@@ -159,10 +161,129 @@ func New(config *Config) (*Client, error) {
 	client.getWebsocket = func(url string) (clientWebsocket, error) {
 		return getWebsocket(transport, url)
 	}
-
+	client.Requester = &requester{
+		baseURL:   &client.baseURL,
+		userAgent: config.UserAgent,
+		client:    &http.Client{Transport: transport},
+		maintenanceHandler: func(err error) {
+			client.maintenance = err
+		},
+		warningHandler: func(count int, timestamp time.Time) {
+			client.warningCount = count
+			client.warningTimestamp = timestamp
+		},
+	}
 	return client, nil
 }
 
+type RequestOptions struct {
+	QueryString url.Values
+	Headers     map[string]string
+	Body        io.Reader
+	Output      interface{}
+}
+
+type ResponseInfo struct {
+	Type     string
+	ChangeID string
+	Result   json.RawMessage
+}
+
+type Requester interface {
+	Request(ctx context.Context, verb, path string, opts *RequestOptions) (*ResponseInfo, error)
+}
+
+type requester struct {
+	baseURL            *url.URL
+	userAgent          string
+	client             *http.Client
+	maintenanceHandler func(error)
+	warningHandler     func(count int, timestamp time.Time)
+}
+
+func (r *requester) Request(ctx context.Context, verb, path string, opts *RequestOptions) (*ResponseInfo, error) {
+	url := *r.baseURL
+	url.Path = pathpkg.Join(url.Path, path)
+	url.RawQuery = opts.QueryString.Encode()
+
+	req, err := http.NewRequestWithContext(ctx, verb, url.String(), opts.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	if r.userAgent != "" {
+		req.Header.Set("User-Agent", r.userAgent)
+	}
+	for h, v := range opts.Headers {
+		req.Header.Set(h, v)
+	}
+
+	retry := time.NewTicker(doRetry)
+	defer retry.Stop()
+	timeout := time.After(doTimeout)
+
+	var res *http.Response
+	for {
+		res, err = r.client.Do(req)
+		if err == nil || verb != "GET" {
+			break
+		}
+		select {
+		case <-retry.C:
+			continue
+		case <-timeout:
+		}
+		break
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	defer res.Body.Close()
+
+	var rsp response
+	d := json.NewDecoder(res.Body)
+	if decodeErr := d.Decode(&rsp); decodeErr != nil {
+		if buf, ioErr := ioutil.ReadAll(d.Buffered()); ioErr != nil {
+			return nil, fmt.Errorf("cannot read buffered response body: %w", ioErr)
+		} else {
+			return nil, fmt.Errorf("cannot decode %q: %w", buf, decodeErr)
+		}
+	}
+	if r.maintenanceHandler != nil {
+		r.maintenanceHandler(rsp.Maintenance)
+	}
+	if err := rsp.err(nil); err != nil {
+		return nil, err
+	}
+
+	if rsp.Type == "sync" {
```

Comment:
It seems to me the caller should be able to say "I'm expecting a `sync` (or `async`) request". If you're expecting an `async` but get a `sync`, I think this will succeed, but it shouldn't.

Whether that means they pass in an expected request type, or we have two different functions again, I'm not sure.

Reply:
I will look into this. Thank you!